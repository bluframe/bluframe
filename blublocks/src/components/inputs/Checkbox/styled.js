// @flow
/* eslint-disable no-ternary */

import styled, { type StyledComponent } from "styled-components"
import type { Theme } from "@bluframe/blublocks"

export const Wrapper: StyledComponent<{||}, Theme, HTMLElement> = styled.div`
  align-items: center;
  column-gap: 8px;
  cursor: pointer;
  display: grid;
  grid-template-columns: auto 1fr;
  height: 16px;
  width: 16px;
`

export const Input: StyledComponent<
  {||},
  Theme,
  HTMLElement
> = styled.input.attrs({ type: "checkbox" })`
  border: 0;
  clip: rect(0 0 0 0);
  clippath: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
`

export const StyledCheckbox: StyledComponent<
  {| checked: boolean |},
  Theme,
  HTMLElement
> = styled.div`
  background: ${({ checked, theme }) =>
    checked ? theme.palette.primary.main : theme.palette.primary.contrast};
  border: ${({ checked, theme }) =>
    checked ? "none" : `1px solid ${theme.palette.primary.main}`};
  border-radius: 3px;
  display: inline-block;
  height: 16px;
  transition: all 150ms;
  width: 16px;
`

export const Icon: StyledComponent<{||}, Theme, any> = styled.svg`
  fill: none;
  stroke: white;
  stroke-width: 2px;
`

export const Polyline: StyledComponent<{||}, Theme, any> = styled.polyline``

export const Label: StyledComponent<
  {||},
  Theme,
  HTMLLabelElement
> = styled.label`
  color: ${({ theme }) => theme.palette.text.main};
  cursor: pointer;
  font-size: 16px;
`
