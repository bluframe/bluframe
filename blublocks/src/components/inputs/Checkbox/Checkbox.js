// @flow
/* eslint-disable no-ternary */

import { Icon, Input, Label, Polyline, StyledCheckbox, Wrapper } from "./styled"
import type { ComponentProps } from "."
import React from "react"

const Checkbox = ({
  inputId,
  isChecked,
  label,
  labelId,
  name,
  onChange,
  value
}: ComponentProps): React$Node => (
  <Wrapper onClick={onChange}>
    <Input defaultChecked={isChecked} name={name} value={value} />

    <StyledCheckbox
      aria-checked={isChecked}
      aria-labelledby={labelId}
      checked={isChecked}
      id={inputId}
      role="checkbox"
    >
      <Icon viewBox="0 0 24 24">
        <Polyline points="20 6 9 17 4 12" />
      </Icon>
    </StyledCheckbox>

    <Label id={labelId}>{label}</Label>
  </Wrapper>
)

export default Checkbox
