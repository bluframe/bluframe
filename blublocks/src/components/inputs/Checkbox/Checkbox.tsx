/* eslint-disable no-ternary */

import { Icon, Input, Label, Polyline, StyledCheckbox, Wrapper } from "./styled"
import { ComponentProps } from "."

const Checkbox = ({
  inputId,
  isChecked,
  label,
  labelId,
  name,
  onClick,
  value
}: ComponentProps) => (
  <Wrapper onClick={onClick}>
    <Input defaultChecked={isChecked} name={name} value={value} />

    <StyledCheckbox
      aria-checked={isChecked}
      aria-labelledby={labelId}
      checked={isChecked}
      id={inputId}
      role="checkbox"
    >
      <Icon viewBox="0 0 24 24">
        <Polyline points="20 6 9 17 4 12" />
      </Icon>
    </StyledCheckbox>

    <Label id={labelId}>{label}</Label>
  </Wrapper>
)

export default Checkbox
