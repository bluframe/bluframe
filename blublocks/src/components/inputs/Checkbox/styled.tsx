/* eslint-disable no-ternary */

import { Theme } from "styles/theme"
import styled from "styled-components"

export const Wrapper = styled.div`
  align-items: center;
  column-gap: 8px;
  cursor: pointer;
  display: grid;
  grid-template-columns: auto 1fr;
`

export const Input = styled.input.attrs({ type: "checkbox" })`
  border: 0;
  clip: rect(0 0 0 0);
  clippath: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
`

export const StyledCheckbox = styled.div`
  background: ${({ checked, theme }: { checked: boolean; theme: Theme }) =>
    checked ? theme.palette.primary.main : theme.palette.primary.contrast};
  border: ${({ checked, theme }: { checked: boolean; theme: Theme }) =>
    checked ? "none" : `1px solid ${theme.palette.primary.main}`};
  border-radius: 3px;
  display: inline-block;
  height: 16px;
  transition: all 150ms;
  width: 16px;
`

export const Icon = styled.svg`
  fill: none;
  stroke: white;
  stroke-width: 2px;
`

export const Polyline = styled.polyline``

export const Label = styled.label`
  color: ${({ theme }: { theme: Theme }) => theme.palette.text.main};
  cursor: pointer;
  font-size: 16px;
`
