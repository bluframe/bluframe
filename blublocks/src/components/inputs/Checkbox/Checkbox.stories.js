// @flow

import Checkbox, { type Props } from "components/inputs/Checkbox"
import React from "react"
import type { Story } from "@storybook/react"
import { action } from "@storybook/addon-actions"

const stories = {
  component: Checkbox,
  title: "Inputs/Checkbox"
}

const Template = (args: Props) => <Checkbox {...args} />

const args = {
  inputId: "apple-checkbox",
  isChecked: true,
  label: "Apple",
  name: "fruit",
  onChange: action("onChange"),
  value: "apple"
}

export const Checked: Story<Props> = Template.bind({})
Checked.args = args

export const Unchecked: Story<Props> = Template.bind({})
Unchecked.args = {
  ...args,
  isChecked: false
}

export default stories
