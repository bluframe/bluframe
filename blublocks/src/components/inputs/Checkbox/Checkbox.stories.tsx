import Checkbox from "components/inputs/Checkbox"
import { StoryObj } from "@storybook/react"
import { action } from "@storybook/addon-actions"

const meta = {
  component: Checkbox,
  title: "Inputs/Checkbox"
}

type Story = StoryObj<typeof Checkbox>

const args = {
  inputId: "apple-checkbox",
  isChecked: true,
  label: "Apple",
  name: "fruit",
  onChange: action("onChange"),
  value: "apple"
}

export const Checked: Story = {
  args
}

export const Unchecked: Story = {
  args: {
    ...args,
    isChecked: false
  }
}

export default meta
