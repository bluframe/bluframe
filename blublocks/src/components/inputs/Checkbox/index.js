// @flow

import Checkbox from "./Checkbox"
import type { CheckboxProps } from "@bluframe/blublocks"
import { prepareComponent } from "@bluframe/grapple"

export type Props = {|
  ...CheckboxProps
|}

export type ComponentProps = {|
  ...Props,
  +labelId?: string
|}

const usePrepareComponent = (props: Props): ComponentProps => {
  const labelId = props.inputId && `${props.inputId}-label`

  const onChange = () => {
    props.onChange(props.value)
  }

  return {
    ...props,
    labelId,
    onChange
  }
}

export default (prepareComponent<Props, ComponentProps>(usePrepareComponent)(
  Checkbox
): React$ComponentType<Props>)
