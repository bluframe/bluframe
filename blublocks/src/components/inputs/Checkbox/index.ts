import React, { ChangeEventHandler } from "react"
import Checkbox from "./Checkbox"
import { prepareComponent } from "@bluframe/grapple"

export type OptionValue = number | string

export interface Props {
  inputId?: string
  isChecked: boolean
  label: React.ReactNode
  name: string
  // eslint-disable-next-line no-unused-vars
  onChange: (value: OptionValue) => void
  value: OptionValue
}

export interface ComponentProps extends Omit<Props, "onChange"> {
  labelId?: string
  onChange: ChangeEventHandler<HTMLInputElement>
}

const usePrepareComponent = (props: Props): ComponentProps => {
  const labelId = props.inputId && `${props.inputId}-label`

  const onChange: ChangeEventHandler<HTMLInputElement> = (event) => {
    event.stopPropagation()

    props.onChange(props.value)
  }

  return {
    ...props,
    labelId,
    onChange
  }
}

export default prepareComponent<Props, ComponentProps>(usePrepareComponent)(
  Checkbox
)
