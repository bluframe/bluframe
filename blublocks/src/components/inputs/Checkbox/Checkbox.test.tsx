import Checkbox from "./Checkbox"
import { ComponentProps } from "."
import React from "react"
import { render } from "tests"

describe("Checkbox", () => {
  const props: ComponentProps = {
    inputId: "apple-checkbox",
    isChecked: true,
    label: "Apple",
    name: "Fruit",
    onClick: jest.fn(),
    value: "apple"
  }

  it("renders", () => {
    const { container } = render(<Checkbox {...props} />)

    expect(container.firstChild).toMatchSnapshot()
  })
})
