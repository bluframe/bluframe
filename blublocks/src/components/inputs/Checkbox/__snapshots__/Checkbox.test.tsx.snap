// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders 1`] = `
<div
  class="sc-bdDqHD BFODt"
>
  <input
    checked=""
    class="sc-gtcySc hwatXx"
    name="Fruit"
    type="checkbox"
    value="apple"
  />
  <div
    aria-checked="true"
    class="sc-dkXqGt eHzDSA"
    id="apple-checkbox"
    role="checkbox"
  >
    <svg
      class="sc-hKVqoC iqYjLd"
      viewBox="0 0 24 24"
    >
      <polyline
        class="sc-eCQicT"
        points="20 6 9 17 4 12"
      />
    </svg>
  </div>
  <label
    class="sc-jSpqns kviUPp"
  >
    Apple
  </label>
</div>
`;
